Implement model for String.equals

codebyzeb opened the pull request:
String.equals currently handled by CBMC, this removes support
within CBMC so that String.equals will be handled by a model instead.
Includes a pointer update to the definition change in the
java-models-library submodule.

codebyzeb committed:
Stop preprocessing String.equals

String.equals currently handled by CBMC, this removes support
within CBMC so that String.equals will be handled by a model instead.
CProverString.equals is provided by JBMC instead of String.equals.

codebyzeb committed:
Update java-models-library submodule

Includes a pointer update to the definition change in the
java-models-library submodule.

codebyzeb commented:
Pointer update to: diffblue/java-models-library#7

tautschnig commented:
I am a nominated reviewer, but I'd hope for @romainbrenguier and @cesaro to weigh in/approve.

romainbrenguier commented on jbmc/regression/strings-smoke-tests/java_value_of_long/test.java:
```diff
@@ -1,5 +1,5 @@
 public class test
-{
+
```
This is probably an involuntary change

romainbrenguier committed:
Use java models in String tests

romainbrenguier committed:
Allow empty arrays in convert_array

martin-cs commented on src/solvers/flattening/boolbv_array.cpp:
```diff
@@ -11,17 +11,16 @@ Author: Daniel Kroening, kroening@kroening.com
 
-  if(width==0)
-    return conversion_failed(expr);
+  if(operands.empty() && width == 0)
```
Is this adding functionality or is this just suppressing an error case?  What is the use-case for this?

romainbrenguier replied:
This is removing a warning in case there is an equality between empty arrays (this was occurring in some of the tests, when I added the java models library to the classpath)

martin-cs replied:
OK but why is the library model comparing empty arrays?  Maybe there is a good reason but equally, it could be a bug, right?

romainbrenguier replied:
I think it's because there are some empty string literals there. I not sure about the exact reason, but I don't see anything wrong with comparing empty arrays, no?

tautschnig replied:
Is an empty string actually an empty array? I'm with Martin, this seems worth investigating. Maybe there are good reasons, but those reasons haven't come up in the last 15 years and thus should be free from doubt and guesswork.

martin-cs replied:
@romainbrenguier : I doubt it is intrinsically harmful but I suspect that quite a bit of the rest of the array handling code was written with the implicit assumption that the arrays were non-empty.  Also, whenever someone removes defensive code in favour of handling a larger set of programs, I like to make sure it isn't a case of "the fire alarms were annoying so we turned them off".  This could quite well be the right thing to do; I'd just like to know why.

peterschrammel replied:
> This could quite well be the right thing to do; I'd just like to know why.
> Is an empty string actually an empty array? 

In a `{ int length; char **data }` representation, what else than `{ .length = 0, .data = &_data }, _data = {}` should an empty string be? Shall data be special-cased to a null pointer to avoid zero-length arrays?

majakusber replied:
Has this been investigated?

tautschnig replied:
Re @peterschrammel's comment: in C, an empty string is `{ '\0' }`, and thus the, e.g., the `string_abstractiont` code distinguishes the "string length" (which might be zero) from the size of the underlying buffer (which should be non-zero). Of course, however, we are talking about a piece of code that should _not_ be specific to C, so all this might not apply.

peterschrammel replied:
Maybe we should ask the question differently: 'What is the error case that above conditional is supposed to detect?'

tautschnig replied:
`boolbv_width` would return `0` for an array of non-constant size. That, however, seems like a bogus case here as a concrete array for a type of unknown size is (at least to me) hard to imagine. So it's probably true that the only reasonable case is that of zero operands. Would any call site of this function fail when receiving an empty array? It's hard to tell, but as of 70532f992716 we are in a much better shape to detect those.

martin-cs reviewed (commented):
I've was asked to review this but there is only one file within my remit so I'm reluctant to approve.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: b06a1f0415c789d7068299ccda5319162a55574c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84267309
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

romainbrenguier committed:
Stop output of useless warning

The thing we warn about is not an error in itself and it's unlikely the
user will understand what this talks about.

romainbrenguier committed:
Relax condition in instantiate

The fact that the array is always accessed at the same index, is not a
necessary condition for the correctness of the instantiate function.

peterschrammel commented on jbmc/regression/jbmc-strings/cprover/CProverString.java:
```diff
@@ -63,6 +63,14 @@ public static StringBuilder deleteCharAt(StringBuilder sb, int index) {
         return sb.deleteCharAt(index);
     }
 
+    public static boolean equals(String a, String b)
```
Why is this required? Can't you use java-models-library instead?

romainbrenguier replied:
Yes that's right, I will remove that part. :eagle:

majakusber commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -1849,8 +1844,8 @@ static exprt instantiate(
-      compute_inverse_function(stream, axiom.univ_var, val, index);
-    implies_exprt instance(
+      compute_inverse_function(axiom.univ_var, val, index);
+    exprt instance = implies_exprt(
```
Why did this change?

romainbrenguier replied:
I removed this change

majakusber commented on jbmc/regression/jbmc-strings/cprover/CProverString.java:
```diff
@@ -63,6 +63,14 @@ public static StringBuilder deleteCharAt(StringBuilder sb, int index) {
         return sb.deleteCharAt(index);
     }
 
+    public static boolean equals(String a, String b)
```
Why is this here in the regression folder? We are loading models library so I thought the method is implemented there now.

romainbrenguier replied:
:eagle:

mgudemann commented on jbmc/regression/jbmc-strings/StringBuilderInsert/test.desc:
```diff
@@ -1,6 +1,6 @@
 Test.class
---function Test.check
+--function Test.check --cp `../../../../scripts/format_classpath.sh . ../../../lib/java-models-library/target/core-models.jar`
 ^EXIT=10$
```
does this work on windows ?

romainbrenguier replied:
I think format_classpath.sh is made specifically to work on linux, mac and windows (unlike using `:` as a separator)

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 29bbb88a41e8a346290ada05c494b96e51a270ff).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84406021
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

majakusber commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -1836,12 +1836,8 @@ static exprt instantiate(
-    "non constant array should always be accessed at the same index");
   exprt::operandst conjuncts;
-  for(const auto &index : indexes)
+  for(const auto &index : find_indexes(axiom.body, str, axiom.univ_var))
```
I'm not sure if I understood this correctly but if there is more than one index via which we access `str`, all of them would get replaced with the same new index `var` which is not correct?

romainbrenguier replied:
If there are several indexes then we get the conjunction of the replacement. 
For instance for a formula `forall q. s[q+x]='a' && s[q]=c` we would get `s[v] = 'a' && s[v-x] = c && s[v+x] = 'a' && s[v] = c`.
I will add this example to the doxygen.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 7a8e8cbc5386e2468c41fc9e6da965595fd5240a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85036174
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 1f004034a0e65210519884660d8a11b9fcc371f2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85385375
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 365d8ca40d9997cbcc485dd1c71f7a06ef57e17c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85407757
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 408cba983848eb5edfbbb529bee7f218bce7e6cc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86245748
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a5cce052b9502bbbf1da6615343aa38129d08f2d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86248971
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

romainbrenguier committed:
Improve documentation of instantiate

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: fa07dc702a3b8dac3c67dc5b69cfc9ead76693da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86257728
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

romainbrenguier committed:
Use java-models-library instead of java_bycode/library

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 9f3d360cb488e6842b61f4f551584a4456e3ad83).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86437726
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

romainbrenguier commented:
submodule update now pointing to master

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 87bd027e8e6423bcd387af5abf2a99ecac8eda76).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86492208
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
